official/coin_build/man/man3/SoSelectOne.3
.TH "SoSelectOne" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SoSelectOne \- The \fBSoSelectOne\fP class is used to select one value from a set of values\&.
.PP
The output field will be the index'th value of the input multivalue field\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/engines/SoSelectOne\&.h>\fP
.PP
Inherits \fBSoEngine\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual \fBSoType\fP \fBgetTypeId\fP (void) const"
.br
.RI "Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. "
.ti -1c
.RI "virtual const \fBSoFieldData\fP * \fBgetFieldData\fP (void) const"
.br
.ti -1c
.RI "virtual const \fBSoEngineOutputData\fP * \fBgetOutputData\fP (void) const"
.br
.ti -1c
.RI "\fBSoSelectOne\fP (\fBSoType\fP inputtype)"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBSoType\fP \fBgetClassTypeId\fP (void)"
.br
.ti -1c
.RI "static void * \fBcreateInstance\fP (void)"
.br
.ti -1c
.RI "static void \fBinitClass\fP (void)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "\fBSoSFInt32\fP \fBindex\fP"
.br
.ti -1c
.RI "\fBSoMField\fP * \fBinput\fP"
.br
.ti -1c
.RI "\fBSoEngineOutput\fP * \fBoutput\fP"
.br
.in -1c
.SS "Static Protected Member Functions"

.in +1c
.ti -1c
.RI "static const \fBSoFieldData\fP ** \fBgetInputDataPtr\fP (void)"
.br
.ti -1c
.RI "static const \fBSoEngineOutputData\fP ** \fBgetOutputDataPtr\fP (void)"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
The \fBSoSelectOne\fP class is used to select one value from a set of values\&.
.PP
The output field will be the index'th value of the input multivalue field\&. 

Note that this engine's output field deviates a little from the 'standard' output mechanism of the majority of engine classes: the \fBSoSelectOne::output\fP is not a permanent \fBSoEngineOutput\fP instance, but a \fIpointer\fP to a \fBSoEngineOutput\fP instance\&. The reason for this is that it is necessary to allocate the output field dynamically to make it match what the \fBSoSelectOne::input\fP is connected to since the type of the \fBSoSelectOne::output\fP always should be the same as the type of the \fBSoSelectOne::input\fP\&.
.PP
The file format for this field is a bit special\&. Note the leading type specification, which is not necessary for most other field types:
.PP
.PP
.nf
SelectOne {
  type <multivaluefieldtype>
  [...fields...]
}
.fi
.PP
 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SoSelectOne::SoSelectOne (\fBSoType\fP inputtype)"
Constructor\&. Sets the type of the input field\&. The input field must be of type \fBSoMField\fP\&. 
.SH "Member Function Documentation"
.PP 
.SS "virtual \fBSoType\fP SoSelectOne::getTypeId (void) const\fC [virtual]\fP"

.PP
Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. Usage example:
.PP
.PP
.nf
void foo(SoNode * node)
{
  if (node->getTypeId() == SoFile::getClassTypeId()) {
    SoFile * filenode = (SoFile *)node;  // safe downward cast, knows the type
  }
}
.fi
.PP
.PP
For application programmers wanting to extend the library with new nodes, engines, nodekits, draggers or others: this method needs to be overridden in \fIall\fP subclasses\&. This is typically done as part of setting up the full type system for extension classes, which is usually accomplished by using the pre-defined macros available through for instance \fBInventor/nodes/SoSubNode\&.h\fP (SO_NODE_INIT_CLASS and SO_NODE_CONSTRUCTOR for node classes), \fBInventor/engines/SoSubEngine\&.h\fP (for engine classes) and so on\&.
.PP
For more information on writing Coin extensions, see the class documentation of the toplevel superclasses for the various class groups\&. 
.PP
Implements \fBSoBase\fP\&.
.SS "virtual const \fBSoFieldData\fP* SoSelectOne::getFieldData (void) const\fC [virtual]\fP"
Returns a pointer to the class-wide field data storage object for this instance\&. If no fields are present, returns \fCNULL\fP\&. 
.PP
Reimplemented from \fBSoFieldContainer\fP\&.
.SS "virtual const \fBSoEngineOutputData\fP* SoSelectOne::getOutputData (void) const\fC [virtual]\fP"
\fIThis API member is considered internal to the library, as it is not likely to be of interest to the application programmer\&.\fP 
.PP
Implements \fBSoEngine\fP\&.
.SH "Member Data Documentation"
.PP 
.SS "\fBSoSFInt32\fP SoSelectOne::index"
Index of the value from the input field which will be put on the output\&. 
.SS "\fBSoMField\fP * SoSelectOne::input"
The multivalue input field which we will select a single value from for our output\&. 
.SS "\fBSoEngineOutput\fP * SoSelectOne::output"
(\fBSoSField\fP) This is the singlevalue field output containing the index'th value of \fBSoSelectOne::input\fP\&.
.PP
The type of the field will of course match the type of the input field, i\&.e\&. if \fBSoSelectOne::input\fP is an \fBSoMFFloat\fP, \fBSoSelectOne::output\fP will be an \fBSoSFFloat\fP etc\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


insanity/coin/src/engines/SoSelectOne.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

/*!
  \class SoSelectOne SoSelectOne.h Inventor/engines/SoSelectOne.h
  \brief The SoSelectOne class is used to select one value from a set of values.
  \ingroup engines

  The output field will be the index'th value of the input multivalue
  field.


  Note that this engine's output field deviates a little from the
  "standard" output mechanism of the majority of engine classes: the
  SoSelectOne::output is not a permanent SoEngineOutput instance, but
  a \e pointer to a SoEngineOutput instance.  The reason for this is
  that it is necessary to allocate the output field dynamically to
  make it match what the SoSelectOne::input is connected to since the
  type of the SoSelectOne::output always should be the same as the
  type of the SoSelectOne::input.


  \ENGINE_TYPELESS_FILEFORMAT

  \verbatim
  SelectOne {
    type <multivaluefieldtype>
    [...fields...]
  }
  \endverbatim
*/

#include <Inventor/engines/SoSelectOne.h>

#include <cstring>

#include <Inventor/SoInput.h>
#include <Inventor/SoOutput.h>
#include <Inventor/engines/SoEngineOutput.h>
#include <Inventor/errors/SoReadError.h>
#include <Inventor/fields/SoFields.h>
#include <Inventor/lists/SoEngineOutputList.h>

#if COIN_DEBUG
#include <Inventor/errors/SoDebugError.h>
#endif // COIN_DEBUG

#include "engines/SoSubEngineP.h"
#include "SbBasicP.h"
#include "coindefs.h"

/*!
  \var SoMField * SoSelectOne::input
  The multivalue input field which we will select a single value from
  for our output.
*/
/*!
  \var SoSFInt32 SoSelectOne::index
  Index of the value from the input field which will be put on the
  output.
*/
/*!
  \var SoEngineOutput * SoSelectOne::output
  (SoSField) This is the singlevalue field output containing the index'th
  value of SoSelectOne::input.

  The type of the field will of course match the type of the input field,
  i.e. if SoSelectOne::input is an SoMFFloat, SoSelectOne::output will
  be an SoSFFloat etc.
*/

// Can't use the standard SO_ENGINE_SOURCE macro, as this engine
// doesn't keep a class-global set of inputs and outputs: we need to
// make an instance of SoFieldData and SoEngineOutputData for every
// instance of the class, since the input and output fields are
// dynamically allocated.
SO_INTERNAL_ENGINE_SOURCE_DYNAMIC_IO(SoSelectOne);

static bool
SoSelectOne_valid_type(SoType t)
{
  return (t.isDerivedFrom(SoMField::getClassTypeId()) &&
          t.canCreateInstance());
}

/*!
  Constructor. Sets the type of the input field. The input field must
  be of type SoMField.
*/
SoSelectOne::SoSelectOne(SoType inputtype)
{
  this->dynamicinput = nullptr;
  this->dynamicoutput = nullptr;
  this->input = nullptr;
  this->output = nullptr;

#if COIN_DEBUG
  if (!SoSelectOne_valid_type(inputtype)) {
    SoDebugError::post("SoSelectOne::SoSelectOne",
                       "invalid type '%s' for input field, "
                       "field must be non-abstract and a multi-value type.",
                       inputtype == SoType::badType() ? "badType" :
                       inputtype.getName().getString());
  }
#endif // COIN_DEBUG

  this->initialize(inputtype);
}

// Default constructor. Leaves engine in invalid state. Should only be
// used from import code or copy code.
SoSelectOne::SoSelectOne(void)
{
  this->dynamicinput = nullptr;
  this->dynamicoutput = nullptr;
  this->input = nullptr;
  this->output = nullptr;
}

// Set up the input and output fields of the engine. This is done from
// either the non-default constructor or the readInstance() import
// code.
void
SoSelectOne::initialize(const SoType inputfieldtype)
{
  assert(this->input == nullptr);
  assert(SoSelectOne_valid_type(inputfieldtype));

  SO_ENGINE_INTERNAL_CONSTRUCTOR(SoSelectOne);
  SO_ENGINE_ADD_INPUT(index, (0));

  // Instead of SO_ENGINE_ADD_INPUT().
  this->input = static_cast<SoMField *>(inputfieldtype.createInstance());
  this->input->setNum(0);
  this->input->setContainer(this);
  this->dynamicinput = new SoFieldData(SoSelectOne::inputdata);
  this->dynamicinput->addField(this, "input", this->input);

  SbString multiname = inputfieldtype.getName().getString();
  // Built-in fields always start with the "MF", but we try to handle
  // user-defined fields aswell.
  const char * ptr = std::strstr(multiname.getString(), "MF");
  assert(ptr != nullptr && "invalid input field type");
  const ptrdiff_t offset = ptr - multiname.getString();
  SbString singlename = (offset == 0) ? SbString("") : multiname.getSubString(0, int(offset - 1));
  singlename += 'S';
  singlename += multiname.getSubString(int(offset + 1));

  SoType outputtype = SoType::fromName(singlename);
  assert(outputtype != SoType::badType() &&
         outputtype.isDerivedFrom(SoSField::getClassTypeId()) &&
         "invalid input field type");

  // Instead of SO_ENGINE_ADD_OUTPUT().
  this->output = new SoEngineOutput;
  this->dynamicoutput = new SoEngineOutputData(SoSelectOne::outputdata);
  this->dynamicoutput->addOutput(this, "output", this->output, outputtype);
  this->output->setContainer(this);
}

// Documented in superclass.
void
SoSelectOne::initClass(void)
{
  SO_ENGINE_INTERNAL_INIT_CLASS(SoSelectOne);
}

SoSelectOne::~SoSelectOne()
{
  delete this->dynamicinput;
  delete this->dynamicoutput;

  delete this->input;
  delete this->output;
}

// Documented in superclass.
void
SoSelectOne::evaluate(void)
{
  int idx = this->index.getValue();

  if (idx == 0 && this->input->getNum() == 0) {
    // Nil is the no-op value (also the default initial value).
    SO_ENGINE_OUTPUT((*output), SoSField, setDirty(false));
  }
  else if (idx >= 0 && idx < this->input->getNum()) {

    // Macro used to generate the right casts for copying field values
#define IF_TYPE(_var_, _fieldtype_) \
    if(_var_ == SoMF##_fieldtype_::getClassTypeId() ) \
    { \
        SO_ENGINE_OUTPUT((*output), SoSF##_fieldtype_, setValue((*(coin_assert_cast<SoMF##_fieldtype_ *>(this->input)))[idx])); \
    }
    // end of macro

    SoType type = this->input->getTypeId();
    IF_TYPE(type,BitMask)
    else IF_TYPE(type,Bool)
    else IF_TYPE(type,Color)
    else IF_TYPE(type,Engine)
    else IF_TYPE(type,Enum)
    else IF_TYPE(type,Float)
    else IF_TYPE(type,Int32)
    else IF_TYPE(type,Matrix)
    else IF_TYPE(type,Name)
    else IF_TYPE(type,Node)
    else IF_TYPE(type,Path)
    else IF_TYPE(type,Plane)
    else IF_TYPE(type,Rotation)
    else IF_TYPE(type,Short)
    else IF_TYPE(type,String)
    else IF_TYPE(type,Time)
    else IF_TYPE(type,UInt32)
    else IF_TYPE(type,UShort)
    else IF_TYPE(type,Vec2f)
    else IF_TYPE(type,Vec3f)
    else IF_TYPE(type,Vec4f)
    else {
      // fall back for user defined types, and built-in types not
      // covered by the above (if any)
      SbString valuestring;
      this->input->get1(idx, valuestring);
      SO_ENGINE_OUTPUT((*output), SoSField, set(valuestring.getString()));
    }
#undef IF_TYPE
  }
#if COIN_DEBUG
  else {
    SoDebugError::post("SoSelectOne::evaluate", "invalid index %d", idx);
  }
#endif // COIN_DEBUG
}

// Documented in superclass.
bool
SoSelectOne::readInstance(SoInput * in, unsigned short flagsarg)
{
  // This code is identical to readInstance() of SoGate and
  // SoConcatenate, so migrate changes.

  SbName tmp;
  if (!in->read(tmp) || tmp != "type") {
    SoReadError::post(in,
                      "\"type\" keyword is missing, erroneous format for "
                      "engine class '%s'.",
                      this->getTypeId().getName().getString());
    return false;
  }
  // need to use an SbString here, because SoInput::read( SbName & )
  // reads in '"MyName"' as is instead of as 'MyName'.
  SbString fieldname;
  if (!in->read(fieldname)) {
    SoReadError::post(in, "Couldn't read input type for engine.");
    return false;
  }
  SoType inputtype = SoType::fromName(fieldname);
  if (!SoSelectOne_valid_type(inputtype)) {
    SoReadError::post(in, "Type \"%s\" for input field is not valid "
                      "(field must be non-abstract and a multi-value type).",
                      fieldname.getString());
    return false;
  }

  this->initialize(inputtype);
  return SoEngine::readInstance(in, flagsarg);
}

// Documented in superclass.
void
SoSelectOne::writeInstance(SoOutput * out)
{
  // This code is identical to writeInstance() of SoGate and
  // SoConcatenate, so migrate changes.

  if (this->writeHeader(out, false, true)) return;

  bool binarywrite = out->isBinary();

  if (!binarywrite) out->indent();
  out->write("type");
  if (!binarywrite) out->write(' ');
  out->write(this->input->getTypeId().getName());
  if (binarywrite) out->write(static_cast<unsigned int>(0));
  else out->write('\n');

  this->getFieldData()->write(out, this);
  this->writeFooter(out);
}

// Documented in superclass.
void
SoSelectOne::copyContents(const SoFieldContainer * from,
                          bool copyconnections)
{
  const SoSelectOne * selectonesrc = coin_assert_cast<const SoSelectOne *>(from);
  if (selectonesrc->input) { this->initialize(selectonesrc->input->getTypeId()); }
  inherited::copyContents(from, copyconnections);
}
